eth/downloader/peer.go

```diff
@@ -305,13 +311,42 @@ func (p *peer) Demote() {
 	}
 }
 
+// Ignore appends a new entity to the set of items (blocks, receipts, states)
+// that a peer is known not to have (i.e. have been requested before). If the
+// set reaches its maximum allowed capacity, items are randomly dropped off.
+func (p *peer) Ignore(hash common.Hash) {
+	p.ignoreLock.Lock()
+	defer p.ignoreLock.Unlock()
+
+	for len(p.ignored) >= maxIgnoredHashes {
+		for drop, _ := range p.ignored {
+			delete(p.ignored, drop)
+			break
+		}
+	}
+	p.ignored[hash] = struct{}{}
+}
+
+// Ignored retrieves whether the hash of a blockchain item is on the peers ignore
+// list or not (i.e. whether we know that the peer does not have it).
```

Comment:
The fact that we need to explain what the 'ignore list' is makes me wonder whether this is the right name.
Consider calling it `failedReq` (and the map `failCache`) or `lacks` (and the map `lacking`).
It should also not be exported.